lib/Sema/TypeCheckMacros.cpp

```diff
@@ -476,6 +476,104 @@ ExpandPeerMacroRequest::evaluate(Evaluator &evaluator, Decl *decl) const {
+///   - Types with `@main` attributes
+///   - Top-level default literal type overrides
+///   - Value decls with names not covered by the macro declaration.
+static void validateMacroExpansion(SourceFile *expansionBuffer,
+                                   MacroDecl *macro,
+                                   ValueDecl *attachedTo,
+                                   MacroRole role) {
+  // Gather macro-introduced names
+  llvm::SmallVector<DeclName, 2> introducedNames;
+  macro->getIntroducedNames(role, attachedTo, introducedNames);
+
+  llvm::SmallDenseSet<DeclName, 2> coversName(introducedNames.begin(),
+                                              introducedNames.end());
+
+  for (auto *decl : expansionBuffer->getTopLevelDecls()) {
+    auto &ctx = decl->getASTContext();
+
+    // Certain macro roles can generate special declarations.
+    if ((isa<AccessorDecl>(decl) && role == MacroRole::Accessor) ||
+        (isa<ExtensionDecl>(decl) && role == MacroRole::Conformance)) {
+      continue;
+    }
+
+    // Diagnose invalid declaration kinds.
+    if (isa<ImportDecl>(decl) ||
+        isa<OperatorDecl>(decl) ||
+        isa<PrecedenceGroupDecl>(decl) ||
+        isa<MacroDecl>(decl) ||
+        isa<ExtensionDecl>(decl)) {
+      decl->diagnose(diag::invalid_decl_in_macro_expansion,
+                     decl->getDescriptiveKind());
+      decl->setInvalid();
+
+      if (auto *extension = dyn_cast<ExtensionDecl>(decl)) {
+        extension->setExtendedNominal(nullptr);
+      }
+
+      continue;
+    }
+
+    // Diagnose `@main` types.
+    if (auto *mainAttr = decl->getAttrs().getAttribute<MainTypeAttr>()) {
+      ctx.Diags.diagnose(mainAttr->getLocation(),
+                         diag::invalid_main_type_in_macro_expansion);
+      mainAttr->setInvalid();
+    }
+
+    // Diagnose default literal type overrides.
+    if (auto *typeAlias = dyn_cast<TypeAliasDecl>(decl)) {
+      auto name = typeAlias->getBaseIdentifier();
+#define EXPRESSIBLE_BY_LITERAL_PROTOCOL_WITH_NAME(_, __, typeName,     \
+                                                  supportsOverride)    \
+      if (supportsOverride && name == makeIdentifier(ctx, typeName)) { \
+        typeAlias->diagnose(diag::literal_type_in_macro_expansion,     \
+                            makeIdentifier(ctx, typeName));            \
+        typeAlias->setInvalid();                                       \
+        continue;                                                      \
+      }
+#include "swift/AST/KnownProtocols.def"
+#undef EXPRESSIBLE_BY_LITERAL_PROTOCOL_WITH_NAME
+    }
+
+    // Diagnose value decls with names not covered by the macro
+    if (auto *value = dyn_cast<ValueDecl>(decl)) {
+      auto baseName = value->getBaseName();
+      if (baseName.isSpecial()) {
+        baseName = ctx.getIdentifier(baseName.userFacingName());
+      }
+
+      // $-prefixed names are unique names. These are always allowed.
+      if (baseName.getIdentifier().hasDollarPrefix()) {
```

Comment:
It's fine for it to be a follow-up, but I think that when there is a dollar prefix, we should go demangle the name and make sure it's one of our unique names.